spicy/src/compiler/codegen/parser-builder.cc

```diff
@@ -1154,6 +1165,16 @@ void ParserBuilder::initializeUnit(const Location& l) {
 void ParserBuilder::finalizeUnit(bool success, const Location& l) {
     const auto& unit = state().unit.get();
 
+    if ( success ) {
+        // We evaluate any "%requires" before running the final "%done" hook
+        // so that (1) that one can rely on the condition, and (2) we keep
+        // running either "%done" or "%error".
+        for ( auto cond : unit.propertyItems("%requires") ) {
+            pushBuilder(builder()->addIf(builder::not_(*cond.expression())),
+                        [&]() { parseError("%requires failed", cond.meta()); });
```

Comment:
This will generate a parser error as soon as any constraint fails (i.e., later constraints will not be validated and not emit additional error messages. WDYT about instead aggregating all failed constraints and emitting the error only after all have been checked. I believe this could be a useful debugging tool, but could have other interactions when e.g., a constraint builds on a previous one (e.g., first check fields are set, then check their values).

Reply:
Would that difference actually be observable? It would fail with a single "parse error" message either way, we can report that only once. If an expression has a side effect, then it would be noticeable but I'd argue it shouldn't. :)

Reply:
> Would that difference actually be observable?
Yes, if the expectation is that the order in which `requires` attributes are declared has significance, my previous example:
```
type X = unit {
    %requires=(self.a? && self.b);     # (1)
    %requires=(self.a + self.b < 42); # (2)
    a: int8 &requires=($$ > 0); # Using field constraints here to make these fields potentially
    b: int8 &requires=($$ > 0); # unset, but similar possible with `switch`.
};
```
The way the code is written suggests that `(1)` would be checked before `(2)` (like it does), so the side effect of `(1)` is that `(2)` can only trigger a parse error, but no exception from accessing an unset optional.

I think with that example I have convinced myself why we should stop checking global constraints as soon as a single one fails 😄